lib/digest/crc.rb

```diff
@@ -93,10 +96,52 @@ def digest_length
     # @param [String] data
     #   The data to update the CRC checksum with.
     #
-    # @abstract
-    #
     def update(data)
-      raise(NotImplementedError,"#{self.class}##{__method__} not implemented")
+      unless @width >= 1
+        raise(NotImplementedError, "incompleted #{self.class} as CRC (expected WIDTH to be 1 or more)")
+      end
+
+      if @table.empty?
+        raise(NotImplementedError, "incompleted #{self.class} as CRC (expected TABLE to be not empty)")
+      end
+
+      if self.class::REFLECT_INPUT.nil?
+        raise(NotImplementedError, "incompleted #{self.class} as CRC (expected REFLECT_INPUT to be not nil)")
+      end
+
+      table = @table
+      crc = @crc
+
+      if self.class::REFLECT_INPUT
+        if @width > 8
+          data.each_byte do |b|
+            crc = table[b ^ (0xff & crc)] ^ (crc >> 8)
```

Comment:
If we trust each element of `@table`, we can make just one more speed improvement:

```diff
-      crc = @crc
+      crc = @crc & ~(-1 << @width)

       if self.class::REFLECT_INPUT
         if @width > 8
           data.each_byte do |b|
-            crc = table[b ^ (0xff & crc)] ^ (crc >> 8)
+            crc = table[b ^ crc] ^ (crc >> 8)
```

Reply:
Sorry, my earlier comment is wrong.
`0xff & ... ` can be removed in three places except above.